File: ios/Classes/PlatformInterface/VideoSource.swift
import Foundation

struct VideoSource {
    let path: String
    let type: VideoSourceType
    let headers: [String: String]

    init?(from map: [String: Any]) {
        guard let path = map["path"] as? String else { return nil }
        self.path = path

        let typeRaw = map["type"] as? String ?? ""
        guard let type = VideoSourceType(rawValue: typeRaw) else { return nil }
        self.type = type
        
        guard let headers = map["headers"] as? [String: String] else {return nil }
        self.headers = headers
    }
}

File: android/src/main/kotlin/me/albemala/native_video_player/platform_interface/VideoSource.kt
package me.albemala.native_video_player.platform_interface

data class VideoSource(
    val path: String,
    val type: VideoSourceType,
    val headers: Map<String, String>
) {
    companion object {
        fun from(map: Map<*, *>): VideoSource? {
            val path = map["path"] as? String
                ?: return null
            val typeValue = map["type"] as? String
                ?: return null
            val headersAny = map["headers"] as? Map<*, *>
                ?: return null
            val headers =
                    headersAny.entries
                    .filter { it.key is String && it.value is String }
                    .associate { (it.key as String) to (it.value as String) }
            val type = VideoSourceType.entries
                .firstOrNull {
                    it.value == typeValue
                }
                ?: return null
            return VideoSource(
                path,
                type,
                headers
            )
        }
    }
}
